helix-term/src/ui/document/cursor.rs

```diff
@@ -0,0 +1,194 @@
+use std::borrow::Cow;
+use std::mem::replace;
+
+use helix_core::{LineEnding, Position, RopeGraphemes, RopeSlice};
+use helix_view::editor;
+use helix_view::graphics::Rect;
+use helix_view::theme::Style;
+
+use crate::ui::document::text_render::StyledGrapheme;
+
+#[derive(Debug)]
+pub struct DocumentCursorConfig {
+    pub max_wrap: usize,
+    pub max_indent_retain: usize,
+}
+
+impl DocumentCursorConfig {
+    pub fn new(editor_config: &editor::Config, viewport: &Rect) -> DocumentCursorConfig {
+        DocumentCursorConfig {
+            // provide a lower limit to ensure wrapping works well for tiny screens (like the picker)
+            max_wrap: editor_config
+                .soft_wrap
+                .max_wrap
+                .min(viewport.width as usize / 4),
+            max_indent_retain: editor_config
+                .soft_wrap
+                .max_indent_retain
+                .min(viewport.width as usize / 4),
+        }
+    }
+}
+
+// fn str_is_whitespace(text: &str) -> bool {
+//     text.chars().next().map_or(false, |c| c.is_whitespace())
+// }
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum WordBoundary {
+    Space,
+    Wrap,
+    Newline,
+}
+
+#[derive(Debug)]
+pub struct DocumentCursor<'a> {
+    pub config: DocumentCursorConfig,
+    char_pos: usize,
+    word_buf: Vec<StyledGrapheme<'a>>,
+    word_width: usize,
+    graphemes: RopeGraphemes<'a>,
+    /// postition within the document
+    pos: Position,
+    highlight_scope: (usize, Style),
+}
+
+impl<'a> DocumentCursor<'a> {
+    pub fn new(
+        text: RopeSlice<'a>,
+        start_line: usize,
+        start_char: usize,
+        editor_config: &editor::Config,
+        viewport: &Rect,
+    ) -> DocumentCursor<'a> {
+        DocumentCursor {
+            char_pos: start_char,
+            word_buf: Vec::with_capacity(if editor_config.soft_wrap.enable {
+                viewport.width as usize
+            } else {
+                64
+            }),
+            word_width: 0,
+            graphemes: RopeGraphemes::new(text),
+            pos: Position {
+                row: start_line,
+                col: 0,
+            },
+            config: DocumentCursorConfig::new(editor_config, viewport),
+            highlight_scope: (0, Style::default()),
+        }
+    }
+
+    // pub fn byte_pos(&self) -> usize {
+    //     self.graphemes.byte_pos()
+    // }
+
+    // pub fn char_pos(&self) -> usize {
+    //     self.char_pos
+    // }
```

Comment:
I want to keep these functions for use in future work but  they are currently not used. and hence generate warnings.